pkg/frontend/querymiddleware/model_extra.go

```diff
@@ -838,247 +842,172 @@ func (m *PrometheusSeriesResponse) Reset()         { *m = PrometheusSeriesRespon
 	}
 	ss := sss[0]
 	if len(ss.Labels) != 1 || ss.Labels[0].Name != "value" {
-		return nil, fmt.Errorf("string sample stream should have exactly one label called value, got %d: %v", len(ss.Labels), ss.Labels)
+		return fmt.Errorf("string sample stream should have exactly one label called value, got %d: %v", len(ss.Labels), ss.Labels)
 	}
-	l := ss.Labels[0]
-
 	if len(ss.Samples) != 1 {
-		return nil, fmt.Errorf("string sample stream should have exactly one sample, got %d", len(ss.Samples))
-	}
-	s := ss.Samples[0]
-
-	return json.Marshal(model.String{Value: l.Value, Timestamp: model.Time(s.TimestampMs)})
-}
-
-func (sss *stringSampleStreams) UnmarshalJSON(b []byte) error {
-	var sv model.String
-	if err := json.Unmarshal(b, &sv); err != nil {
-		return err
+		return fmt.Errorf("string sample stream should have exactly one sample, got %d", len(ss.Samples))
 	}
-	*sss = []SampleStream{{
-		Labels:  []mimirpb.LabelAdapter{{Name: "value", Value: sv.Value}},
-		Samples: []mimirpb.Sample{{TimestampMs: int64(sv.Timestamp)}},
-	}}
 	return nil
 }
 
-type scalarSampleStreams []SampleStream
+// A string sample is written like `[12345678, "hello"]`
+func stringSampleEncode(sss []SampleStream, stream *jsoniter.Stream) {
+	stream.WriteArrayStart()
+	stream.WriteRaw(model.Time(sss[0].Samples[0].TimestampMs).String())
+	stream.WriteMore()
+	stream.WriteString(sss[0].Labels[0].Value)
+	stream.WriteArrayEnd()
+}
 
-func (sss scalarSampleStreams) MarshalJSON() ([]byte, error) {
+func validateScalarSampleStream(sss []SampleStream) error {
 	if len(sss) != 1 {
-		return nil, fmt.Errorf("scalar sample streams should have exactly one stream, got %d", len(sss))
+		return fmt.Errorf("scalar sample streams should have exactly one stream, got %d", len(sss))
 	}
 	ss := sss[0]
 	if len(ss.Samples) != 1 {
-		return nil, fmt.Errorf("scalar sample stream should have exactly one sample, got %d", len(ss.Samples))
+		return fmt.Errorf("scalar sample stream should have exactly one sample, got %d", len(ss.Samples))
 	}
-	s := ss.Samples[0]
-	return json.Marshal(model.Scalar{
-		Timestamp: model.Time(s.TimestampMs),
-		Value:     model.SampleValue(s.Value),
-	})
-}
-
-func (sss *scalarSampleStreams) UnmarshalJSON(b []byte) error {
-	var sv model.Scalar
-	if err := json.Unmarshal(b, &sv); err != nil {
-		return err
-	}
-	*sss = []SampleStream{{
-		Samples: []mimirpb.Sample{{TimestampMs: int64(sv.Timestamp), Value: float64(sv.Value)}},
-	}}
 	return nil
 }
 
-// asVectorSampleStreams converts a slice of SampleStream into a slice of vectorSampleStream.
-// This can be done as vectorSampleStream is defined as a SampleStream.
-func asVectorSampleStreams(ss []SampleStream) []vectorSampleStream {
-	return *(*[]vectorSampleStream)(unsafe.Pointer(&ss))
+// A scalar sample is written like `[12345678, 42]`
+func scalarSampleEncode(sss []SampleStream, stream *jsoniter.Stream) {
+	s := sss[0].Samples[0]
+	stream.WriteArrayStart()
+	stream.WriteRaw(model.Time(s.TimestampMs).String())
```

Comment:
This should really be `jsonutil.MarshalTimestamp`, but that formats numbers differently, e.g. `1.500` instead of `1.5`, and there are tests which check this.